stan/math/rev/fun/squared_distance.hpp

```diff
@@ -130,27 +124,23 @@ class squared_distance_vd_vari : public vari {
 };
 }  // namespace internal
 
-template <int R1, int C1, int R2, int C2>
-inline var squared_distance(const Eigen::Matrix<var, R1, C1>& v1,
-                            const Eigen::Matrix<var, R2, C2>& v2) {
-  check_vector("squared_distance", "v1", v1);
-  check_vector("squared_distance", "v2", v2);
+template <typename T1, typename T2,
+          require_all_eigen_vector_vt<is_var, T1, T2>* = nullptr>
+inline var squared_distance(const T1& v1, const T2& v2) {
   check_matching_sizes("squared_distance", "v1", v1, "v2", v2);
   return var(new internal::squared_distance_vv_vari(v1, v2));
 }
-template <int R1, int C1, int R2, int C2>
-inline var squared_distance(const Eigen::Matrix<var, R1, C1>& v1,
-                            const Eigen::Matrix<double, R2, C2>& v2) {
-  check_vector("squared_distance", "v1", v1);
-  check_vector("squared_distance", "v2", v2);
+template <typename T1, typename T2,
+          require_eigen_vector_vt<is_var, T1>* = nullptr,
+          require_eigen_vector_vt<std::is_arithmetic, T2>* = nullptr>
+inline var squared_distance(const T1& v1, const T2& v2) {
```

Comment:
[Optional] For something like this I think `EigenVar` and `EigenArith` would be useful names

Reply:
done